### PR TITLE
Added IMU measurement input support.

### PR DESCRIPTION
**Pull request by lukes3315**
# New Features
- Added new `IMUInput` message definition

# Fixes
- Corrected `RAW_IMU_OUTPUT` enum name in Python

### CODE REVIEW

File: src/point_one/fusion_engine/messages/defs.h [hunk 1]
```diff
@@ -50,6 +50,8 @@ enum class MessageType : uint16_t {
   RAW_HEADING_OUTPUT = 11001, ///< @ref RawHeadingOutput
   RAW_IMU_OUTPUT = 11002, ///< @ref RawIMUOutput
   HEADING_OUTPUT = 11003, ///< @ref HeadingOutput
+  IMU_INPUT = 11004,
+  RAW_IMU_INPUT = 11005,
```
Comment (adamshapiro0) [inline src/point_one/fusion_engine/messages/defs.h:54]: Remove

File: src/point_one/fusion_engine/messages/defs.h [hunk 2]
```diff
@@ -50,6 +50,8 @@ enum class MessageType : uint16_t {
   RAW_HEADING_OUTPUT = 11001, ///< @ref RawHeadingOutput
   RAW_IMU_OUTPUT = 11002, ///< @ref RawIMUOutput
   HEADING_OUTPUT = 11003, ///< @ref HeadingOutput
+  IMU_INPUT = 11004,
```
[the remark this answers is not in the document]
Comment (adamshapiro0) [reply src/point_one/fusion_engine/messages/defs.h:53]: Also, need to add a case for new `IMU_INPUT` enum below `HEADING_OUTPUT` in the `to_string()` function below

File: src/point_one/fusion_engine/messages/measurements.h [hunk 3]
```diff
@@ -316,6 +316,91 @@ struct P1_ALIGNAS(4) RawIMUOutput : public MessagePayload {
   int32_t gyro[3] = {INT32_MAX, INT32_MAX, INT32_MAX};
 };
 
+/**
+ * @brief IMU sensor measurement input  with calibration and corrections applied
```
Comment (adamshapiro0) [inline src/point_one/fusion_engine/messages/measurements.h:320]: Remove "with ..." -- that only applies to outputs.

File: src/point_one/fusion_engine/messages/measurements.h [hunk 4]
```diff
@@ -316,6 +316,91 @@ struct P1_ALIGNAS(4) RawIMUOutput : public MessagePayload {
...
+ *        (@ref MessageType::IMU_INPUT, version 1.0).
+ * @ingroup measurement_messages
+ *
+ * This message is an input  from the device containing IMU acceleration and
```
Comment (adamshapiro0) [inline src/point_one/fusion_engine/messages/measurements.h:324]: This paragraph applies to outputs, not inputs.

> This message is an input to the device containing IMU acceleration and rotation rate measurements.

File: src/point_one/fusion_engine/messages/measurements.h [hunk 5]
```diff
@@ -316,6 +316,91 @@ struct P1_ALIGNAS(4) RawIMUOutput : public MessagePayload {
...
+ * rotation rate measurements. The measurements been corrected for biases and
+ * scale factors, and have been rotated into the vehicle body frame from the
+ * original IMU orientation, including calibrated mounting error estimates.
+ *
+ * See also @ref RawIMUInput.
```
Comment (adamshapiro0) [inline src/point_one/fusion_engine/messages/measurements.h:329]: Remove

File: src/point_one/fusion_engine/messages/measurements.h [hunk 6]
```diff
@@ -316,6 +316,91 @@ struct P1_ALIGNAS(4) RawIMUOutput : public MessagePayload {
...
+ */
+struct P1_ALIGNAS(4) IMUInput : public MessagePayload {
+  static constexpr MessageType MESSAGE_TYPE = MessageType::IMU_INPUT;
+  static constexpr uint8_t MESSAGE_VERSION = 0;
+
+  /** The time of the measurement, in P1 time (beginning at power-on). */
+  Timestamp p1_time;
```
Comment (adamshapiro0) [inline src/point_one/fusion_engine/messages/measurements.h:336]: Input classes must use `MeasurementDetails`. Please see `WheelSpeedInput` as an example.

File: src/point_one/fusion_engine/messages/measurements.h [hunk 7]
```diff
@@ -316,6 +316,91 @@ struct P1_ALIGNAS(4) RawIMUOutput : public MessagePayload {
...
+
+  /**
+   * Corrected vehicle x/y/z rate of rotation standard deviation (in
+   * radians/second), resolved in the body frame.
+   */
+  double gyro_std_rps[3] = {NAN, NAN, NAN};
+};
+
+/**
+ * @brief Raw (uncorrected) IMU sensor measurement input (@ref
```
Comment (adamshapiro0) [inline src/point_one/fusion_engine/messages/measurements.h:364]: All inputs are uncorrected, so there is no "raw" vs "not" for inputs. This class is not needed.

File: src/point_one/fusion_engine/messages/measurements.h [hunk 8]
```diff
@@ -316,6 +316,91 @@ struct P1_ALIGNAS(4) RawIMUOutput : public MessagePayload {
...
+          MessageType::RAW_IMU_INPUT, version 1.0).
+ * @ingroup measurement_messages
+ *
+ * This message is an input from the device containing raw IMU acceleration and
+ * rotation rate measurements. These measurements come directly from the sensor,
+ * and do not have any corrections or calibration applied.
+ *
+ * See also @ref IMUInput.
```
Comment (adamshapiro0) [inline src/point_one/fusion_engine/messages/measurements.h:372]: Remove

File: src/point_one/fusion_engine/messages/measurements.h [hunk 9]
```diff
@@ -316,6 +316,91 @@ struct P1_ALIGNAS(4) RawIMUOutput : public MessagePayload {
...
+ * @ingroup measurement_messages
+ *
+ * This message is an input  from the device containing IMU acceleration and
+ * rotation rate measurements. The measurements been corrected for biases and
+ * scale factors, and have been rotated into the vehicle body frame from the
+ * original IMU orientation, including calibrated mounting error estimates.
+ *
+ * See also @ref RawIMUInput.
+ */
+struct P1_ALIGNAS(4) IMUInput : public MessagePayload {
```
Comment (adamshapiro0) [inline src/point_one/fusion_engine/messages/measurements.h:331]: See below - let's remove this class in favor of the raw format below, and rename that one.

File: python/fusion_engine_client/messages/measurements.py [hunk 10]
```diff
@@ -97,7 +97,144 @@ class RawIMUOutput(MessagePayload):
...
+        if item == 'p1_time':
+            return self.details.p1_time
+        else:
+            return super().__getattr__(item)
+
+    def __str__(self):
+        return construct_message_to_string(message=self, construct=self.Construct, title='Raw IMU Output',
+                                           fields=['details', 'accel_mps2', 'gyro_rps', 'temperature_degc'])
+
+class IMUInput(MessagePayload):
```
Comment (adamshapiro0) [inline python/fusion_engine_client/messages/measurements.py:153]: Remove this class, keep raw input class below but rename to `IMUInput` (`IMU_INPUT` enum)

File: python/fusion_engine_client/messages/measurements.py [hunk 11]
```diff
@@ -97,7 +97,144 @@ class RawIMUOutput(MessagePayload):
...
+        return result
+
+    @classmethod
+    def calcsize(cls) -> int:
+        return Timestamp.calcsize() + cls._STRUCT.size
+
+
+class RawIMUInput(MessagePayload):
+    """!
+    @brief Raw (uncorrected) IMU sensor measurement output.
```
Comment (adamshapiro0) [inline python/fusion_engine_client/messages/measurements.py:235]: Update please - this is an input, and remove uncorrected since that's implied

**Review by adamshapiro0 — CHANGES_REQUESTED** [review]
A few changes to make here, nothing major but inputs are always implicitly raw, and should use similar formats to the corresponding raw (uncorrected) output messages.

File: src/point_one/fusion_engine/messages/defs.h (hunk 1 above)
Comment (lukes3315) [reply src/point_one/fusion_engine/messages/defs.h:54]: Remove IMU_INPUT & RAW_IMU_INPUT enums?

File: src/point_one/fusion_engine/messages/measurements.h (hunk 7 above)
Comment (lukes3315) [reply src/point_one/fusion_engine/messages/measurements.h:364]: So basically only keep IMUInput? And omit all references to RawIMUInput?

File: src/point_one/fusion_engine/messages/measurements.h (hunk 9 above)
Comment (lukes3315) [reply src/point_one/fusion_engine/messages/measurements.h:331]: Alright I think that makes sense, I was thinking this was going to make users confused lol.

File: src/point_one/fusion_engine/messages/measurements.h [hunk 12]
```diff
@@ -316,6 +316,46 @@ struct P1_ALIGNAS(4) RawIMUOutput : public MessagePayload {
   int32_t gyro[3] = {INT32_MAX, INT32_MAX, INT32_MAX};
 };
 
+/**
+ * @brief IMU sensor measurement input (@ref MessageType::IMU_INPUT,
+ *        version 1.0).
+ * @ingroup measurement_messages
+ *
+ * This message is an input from the device containing raw IMU acceleration and
```
Comment (adamshapiro0) [inline src/point_one/fusion_engine/messages/measurements.h:324]: Input _to_ the device

File: src/point_one/fusion_engine/messages/measurements.h [hunk 13]
```diff
@@ -316,6 +316,46 @@ struct P1_ALIGNAS(4) RawIMUOutput : public MessagePayload {
...
+ * rotation rate measurements.
+ *
+ * See also @ref IMUInput.
```
Comment (adamshapiro0) [inline src/point_one/fusion_engine/messages/measurements.h:327]: This reference is just to the same class. I assume this should reference `@ref IMUOutput` instead?

File: src/point_one/fusion_engine/messages/measurements.h [hunk 14]
```diff
@@ -316,6 +316,46 @@ struct P1_ALIGNAS(4) RawIMUOutput : public MessagePayload {
...
+ */
+struct P1_ALIGNAS(4) IMUInput : public MessagePayload {
```
Comment (adamshapiro0) [inline src/point_one/fusion_engine/messages/measurements.h:329]: For consistency, we've been ordering the classes input, output, raw output. Please move this class up above `IMUOutput` to match the other measurement types.

**Pull request merged**